src/node/node_state.h

```diff
@@ -1317,96 +1317,118 @@ namespace ccf
       // inform the host of any nodes that no longer need to be tracked.
       network.nodes.set_global_hook(
         [this](kv::Version version, const Nodes::Write& w) {
-          for (auto& [node_id, ni] : w)
+          for (const auto& [node_id, opt_ni] : w)
           {
-            if (ni.value.status == NodeStatus::RETIRED)
+            if (!opt_ni.has_value())
+            {
+              throw std::logic_error("Unexpected: removal from nodes table");
+            }
+
+            const auto& ni = opt_ni.value();
+            if (ni.status == NodeStatus::RETIRED)
+            {
               remove_node(node_id);
+            }
           }
         });
 
       network.service.set_global_hook(
         [this](kv::Version version, const Service::Write& w) {
-          if (w.at(0).value.status == ServiceStatus::OPEN)
+          const auto it = w.find(0);
+          if (it == w.end() || !it->second.has_value())
+          {
+            throw std::logic_error(
+              "Unexpected: write to service table does not include key 0");
+          }
+
+          if (it->second.value().status == ServiceStatus::OPEN)
           {
             this->consensus->set_f(1);
             open_user_frontend();
             LOG_INFO_FMT("Network is OPEN, now accepting user transactions");
           }
         });
 
-      network.secrets.set_local_hook(
-        [this](kv::Version version, const Secrets::Write& w) {
-          bool has_secrets = false;
-          std::list<LedgerSecrets::VersionedLedgerSecret> restored_secrets;
+      network.secrets.set_local_hook([this](
+                                       kv::Version version,
+                                       const Secrets::Write& w) {
+        bool has_secrets = false;
+        std::list<LedgerSecrets::VersionedLedgerSecret> restored_secrets;
 
-          for (auto& [v, secret_set] : w)
+        for (const auto& [v, opt_secret_set] : w)
+        {
+          if (!opt_secret_set.has_value())
           {
-            for (auto& encrypted_secret_for_node : secret_set.value.secrets)
+            throw std::logic_error("Unexpected: removal from secrets table");
```

Comment:
Not likely this'll happen, but if it does there's a small chance we may be happy to have the version.